Preemption algorithm described (#1988)
kubernetes-sigs/kueue: site/content/en/docs/concepts/cluster_queue.md

  - `LowerOrNewerEqualPriority`: only preempt Workloads in the ClusterQueue that either have a lower priority than the pending workload or equal priority and are newer than the pending workload.

Note that an incoming Workload can preempt Workloads both within the
ClusterQueue and the cohort.

Kueue implements heuristics to preempt as few Workloads as possible.
Below we present a more detailed description of the algorithm.

### Preemption Algorithm overview

> An incoming Workload, which does not fit within the free capacity, is eligible

Reviewer: ```suggestion
An incoming Workload, which does not fit within the unused quota, is eligible
```

> for preemptions, within a flavor of the target queue, when one of the following

Reviewer: ```suggestion
to issue preemptions when one of the following
```

for simplicity

is true:
- the requests of the Workload are below the flavor's nominal quota, or
- `borrowWithinCohort` is enabled.

#### Candidates

The list of preemption candidates is compiled from Workloads within the Cluster
Queue satisfying the `withinClusterQueue` policy, and Workloads within the
cohort which satisfy the `reclaimWithinCohort` policy.

The list of candidates is sorted based on the following preference checks for
tie-breaking:

> - Workloads from other borrowing queues in the cohort,

Reviewer: ```suggestion
- Workloads from borrowing queues in the cohort,
```

- Workloads with the lowest priority,
- Workloads which got admitted the most recently.

#### Targets

> The algorithm qualifies the candidates as targets using the following heuristics:

Reviewer: ```suggestion
The algorithm qualifies the candidates as preemption targets using the following heuristics:
```

1. If all candidates belong to the target queue, then Kueue greedily

> qualifies candidates until the incoming Workload can fit, allowing for borrowing.

Reviewer: instead of `allowing borrowing`, maybe we can be more explicit.

And then you can refer to "allowing/disallowing borrowing" freely

```
allowing the usage of the ClusterQueue to be above the nominal quota, up to the `borrowingLimit`.
This is called "borrowing".
```

Author: Done along the lines. PTAL.

2. If `borrowWithinCohort` is enabled, then Kueue greedily qualifies
candidates (respecting the `borrowWithinCohort.maxPriorityThreshold` threshold),
until the incoming Workload can fit, allowing for borrowing.

3. If the current usage of the target queue is below nominal quota, then
Kueue greedily qualifies the candidates, until the incoming workload can fit
without borrowing.

4. Kueue tries to greedily qualifies a subset of candidates which belong to the
target Cluster Queue, until the incoming Workload can fit, allowing for borrowing.

The last step of the algorithm is to optimize the set of targets. For this

> purpose Kueue greedily traverses the list of initial targets (in reverse) and

Reviewer: ```suggestion
purpose Kueue greedily traverses the list of initial targets in reverse and
```

> removes these without which the incoming Workload still can be admitted.

Reviewer: ```suggestion
removes them from the list of targets if the incoming Workload still can be admitted when they are accounted back for quota usage.
```

Note that an incoming Workload can preempt Workloads both within the
ClusterQueue and the cohort.

### Preemption Algorithm overview

An incoming Workload, which does not fit within the free capacity, is eligible
for preemptions, within a flavor of the target queue, when one of the following
is true:
- the requests of the Workload are below the flavor's nominal quota, or
- `borrowWithinCohort` is enabled.

Kueue implements heuristics to preempt as few Workloads as possible, preferring
Workloads with these characteristics:

- Workloads belonging to ClusterQueues that are borrowing quota.
- Workloads with the lowest priority.
- Workloads that have been admitted more recently.

Below we present a more detailed description of the algorithm.

#### Candidates

The list of preemption candidates is compiled from Workloads within the Cluster
Queue satisfying the `withinClusterQueue` policy, and Workloads within the
cohort which satisfy the `reclaimWithinCohort` policy.

The list of candidates is sorted based on the following preference checks for
tie-breaking:
- Workloads from other borrowing queues in the cohort,
- Workloads with the lowest priority,

> - Workloads which got admitted the most recently.

Reviewer 2: We should sync the list at L438 with this, and maybe louse one of them.

Just removing L438-L441 can be the best.

Author: Indeed, let me tentatively drop  L438-L441, and moved up as intro for the "Preemption Algorithm overview" section.

- Workloads with the lowest priority.

> - Workloads that have been admitted more recently.

Author: Moved it below. PTAL.

### Preemption Algorithm

> The preemption algorithm has the following stages.

Author: I think `preemptor` and `victim` makes sense, as they used for preemption of pods. I'm going to update using this vocab.

> The preemption algorithm has the following stages.

Reviewer 2: `victim` is way too dark, being evicted is not such a tragedy, and it will add yet another terminology in the context of preemption

> The preemption algorithm has the following stages.

Reviewer: That word is not listed here https://inclusivenaming.org/word-lists/

However, it definitely doesn't sound appropriate. You can use `preemptee` as long as you define it somewhere.

> The preemption algorithm has the following stages.

Reviewer: `target` might work better.

> The preemption algorithm has the following stages.

Author: Victim is used in the user-facing docs for pods, so was reviewed by tech writers I assume: https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/.

However, if you feel it is not good in the context here, then I like target, which I actually use in the text currently.

> The preemption algorithm has the following stages.

Author: I have rephrased the description using "incoming Workload" (as it was already used in other places in the "Preemption" section of the page). The end result I call "targets", the sorted interim list I call "candidates". PTAL.

#### Eligibility

> A workload is eligible for preemptions if it is eligible for preemptions for the

Reviewer 2: `A workload is eligible for preemptions if it is eligible for preemptions` sounds strange.

Author: Right, I have rephrased the description. PTAL.

selected resource flavor
(see [FlavorFungibility](/docs/concepts/cluster_queue/#flavorfungibility)).

A workload is only eligible to do preemptions in a given resource flavor if it
fits fully within the nominal quota of the flavor. Alternatively, if
 `borrowWithinCohort` is enabled, then any workloads requesting within the

>  cohort limits are eligible for preemptions.

Author: I have rephrased the section. PTAL. Regarding the reference to the API, we could, but the API is already described just above the new section.

#### Candidates

The list of peremption candidates is compiled as the list of workloads within
the cluster queue, satisfying the `withinClusterQueue` policy, and workloads
within the cohort which satisfy the `reclaimWithinCohort` policy.

The list of candidates is sorted based on the following preference checks for
tie-breaking:
- workloads from other cluster queues
- workloads with the lowest priority

> - workloads which run the shortest

Author: Right, updated. PTAL

#### Targets

The algorithm to qualify candidates as targets does the following steps:

1. If all candidates belong to the target queue, then Kueue gradually qualifies
candidates as targets until the workload can fit. The end result utilization
might be the above nominal.

2. If `borrowWithinCohort` is enabled, then all candidates are considered in
order, respecting the `borrowWithinCohort.maxPriorityThreshold` threshold, until
the workload can fit. The end result utilization might be above the nominal
quota.

3. If existing workloads are under nominal, then all candidates are considered
in order. However, the final utilization cannot exceed the nominal quota.

4. Only a subset of candidates from the target queue is considered, but the end

> result utilization is allowed to be above nominal.

Author: Thanks for the heads up, I'm going to think how to rephrase them. In the meanwhile, they are descriptions of this steps in code: https://github.com/kubernetes-sigs/kueue/blob/ebecc565094f0555903a037502c05e99c081d470/pkg/scheduler/preemption/preemption.go#L122-L130.

> result utilization is allowed to be above nominal.

Author: I have rephrased them. PTAL.
